src/com/amazon/ion/impl/IonReaderBinaryRawX.java

```diff
@@ -1181,81 +1202,106 @@ protected final Timestamp readTimestamp(int len) throws IOException
-        // chars even after utf8 to utf16 decoding nonsense
-        // the char array is way faster than using string buffer
-        char[] chars = new char[len];
-        int    c, ii = 0;
-        int    save_limit = NO_LIMIT;
-        if (_local_remaining != NO_LIMIT) {
-            save_limit = _local_remaining - len;
+        // If the string we're reading is small enough to fit in our reusable buffer, we can avoid the overhead
+        // of looping and bounds checking.
+        if (numberOfBytes < utf8InputBuffer.array().length) {
+            return readStringWithReusableBuffer(numberOfBytes);
         }
-        _local_remaining = len;
-        while (!isEOF()) {
-            c = readUnicodeScalar();
-            if (c < 0) throwUnexpectedEOFException();
-            if (c < 0x10000) {
-                chars[ii++] = (char)c;
+
+        // Otherwise, allocate a one-off decoding buffer that's large enough to hold the string
+        // and prepare to decode the string in chunks.
+        CharBuffer decodingBuffer = CharBuffer.allocate(numberOfBytes);
+        utf8CharsetDecoder.reset();
+
+        int save_limit = NO_LIMIT;
+        if (_local_remaining != NO_LIMIT) {
+            save_limit = _local_remaining - numberOfBytes;
+        }
+        _local_remaining = numberOfBytes;
+
+        // The following loop will:
+        // * Fill the input buffer with utf8 bytes
+        // * Write decoded chars to the decoding buffer
+        // * Move any remaining partial character bytes to the front of the buffer
+        // * Repeat until the requested number of bytes have been decoded.
+        // * Create a new String object from the contents of the decoding buffer.
+        int totalBytesRead = 0;
+        int carryoverBytes = 0;
+        while (totalBytesRead < numberOfBytes) {
+
+            int bytesRemaining = numberOfBytes - totalBytesRead;
+            // When decoding, it's possible to have 'carryover' bytes left in the buffer which represent
+            // a partial unicode character. We need to leave these in the buffer so that we can complete the
+            // partial character in the next call to read().
+            int capacityRemaining = utf8InputBuffer.array().length - carryoverBytes;
+            int bytesToRead = Math.min(bytesRemaining, capacityRemaining);
+
+            int bytesRead = read(utf8InputBuffer.array(), carryoverBytes, bytesToRead);
+            if (bytesRead <= 0) {
+                // UnifiedInputStreamX doesn't adhere to InputStream's API. See the comments on
+                // UnifiedInputStreamX#read() for more information.
+                throwUnexpectedEOFException();
             }
-            else { // when c is >= 0x10000 we need surrogate encoding
-                chars[ii++] = (char)_Private_IonConstants.makeHighSurrogate(c);
-                chars[ii++] = (char)_Private_IonConstants.makeLowSurrogate(c);
+
+            totalBytesRead += bytesRead;
+
+            utf8InputBuffer.position(0);
```

Comment:
it'd be nice to avoid this as well as the `System.arraycopy` below (an exceptional case, i suspect).  i presume this is due to the signature of `read()` accepting an array rather than a `Buffer`--is that correct?

Reply:
> i presume this is due to the signature of read() accepting an array rather than a Buffer--is that correct?

That's correct. Unfortunately, the API mismatch between `InputStream` (which operates on `byte[]`s) and `CharsetDecoder` (which operates on `ByteBuffer`s and `CharBuffer`s) requires a fair amount of complexity to work around without allocating new objects each time. A simpler version of this code that used `CharBuffer.wrap()` was markedly slower.

> it'd be nice to avoid this as well as the System.arraycopy (an exceptional case, i suspect)

It is pretty rare, yes.